docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,359 @@
+3. We would like to be able to track timing metrics for how quickly we can retrieve cat facts.
+
+### Initial setup
+
+``` haskell
+{-# LANGUAGE ExistentialQuantification, DeriveAnyClass, DeriveFunctor,
+DeriveGeneric, DerivingStrategies, FlexibleContexts, FlexibleInstances,
+GeneralizedNewtypeDeriving, OverloadedStrings, MultiParamTypeClasses,
+RankNTypes, TypeApplications, TypeOperators, UndecidableInstances #-}
+module CatFacts
+    ( main
+    ) where
+-- from base
+import Control.Applicative
+import Control.Exception (throwIO)
+import GHC.Generics (Generic1)
+-- from fused-effects
+import Control.Algebra
+import Control.Carrier.Reader
+import Control.Carrier.Error.Either
+import Control.Carrier.Interpret
+-- from transformers
+import Control.Monad.IO.Class
+-- From aeson
+import Data.Aeson
+-- From bytestring
+import qualified Data.ByteString.Char8 as B
+import qualified Data.ByteString.Lazy as L
+-- From time
+import Data.Time.Clock
+-- From http-client
+import qualified Network.HTTP.Client as HTTP
+import Network.HTTP.Client.Internal (Response(..), ResponseClose(..))
+-- From http-client-tls
+import qualified Network.HTTP.Client.TLS as HTTP
+-- From http-status
+import Network.HTTP.Types.Header
+import Network.HTTP.Types.Status
+import Network.HTTP.Types.Version
+```
+
+Since one of the best parts about effects is being able to think at a domain language level,
+let's start with defining the desired data which we wish to retrieve and an interface that
+feels natural to work with:
+
+``` haskell
+-- | The basic fact that we will retrieve.
+newtype CatFact = CatFact
+  { catFact :: String
+  } deriving (Show)
+
+instance FromJSON CatFact where
+  parseJSON = withObject "CatFact" (\o -> CatFact <$> o .: "text")
+
+-- | Our high level effect type that will be able to target different data sources.
+data CatFactClient m k
+  = ListFacts Int {- ^ Number of facts to fetch -} ([CatFact] -> m k)
+  deriving stock (Functor, Generic1)
+  deriving anyclass (HFunctor, Effect)
+
+listFacts :: Has CatFactClient sig m => Int -> m [CatFact]
+listFacts n = send (ListFacts n pure)
+```
+
+Now that we have our very simple DSL in place, let's think about the underlying API: we know that it's an HTTP-based system, so let's introduce the notion of a handler that is provided a request and hands back an HTTP response.
+
+``` haskell
+data Http m k
+  = SendRequest HTTP.Request (HTTP.Response L.ByteString -> m k)
+  deriving stock (Functor, Generic1)
+  deriving anyclass (HFunctor, Effect)
+
+sendRequest :: Has Http sig m => HTTP.Request -> m (HTTP.Response L.ByteString)
+sendRequest r = send (SendRequest r pure)
+```
+
+The listFacts function provides the ‘what’ of this API, and the sendRequest function provides the ‘how’. In decomposing this problem into a set of effects, each responsible for a single layer of the original problem description, we provide ourselves with a flexible, composable vocabulary rather than a single monolithic action.
```

Comment:
```suggestion
The `listFacts` function provides the ‘what’ of this API, and the sendRequest function provides the ‘how’. In decomposing this problem into a set of effects, each responsible for a single layer of the original problem description, we provide ourselves with a flexible, composable vocabulary rather than a single monolithic action.
```